services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRedirectUrisEnforcerExecutor.java

```diff
@@ -0,0 +1,675 @@
+    public Class<Configuration> getExecutorConfigurationClass() {
+        return Configuration.class;
+    }
+
+    public static class Configuration extends ClientPolicyExecutorConfigurationRepresentation {
+        @JsonProperty(SecureRedirectUrisEnforcerExecutorFactory.ALLOW_PRIVATE_USE_URI_SCHEME)
+        protected boolean allowPrivateUseUriScheme;
+        @JsonProperty(SecureRedirectUrisEnforcerExecutorFactory.ALLOW_IPV4_LOOPBACK_ADDRESS)
+        protected boolean allowIPv4LoopbackAddress;
+        @JsonProperty(SecureRedirectUrisEnforcerExecutorFactory.ALLOW_IPV6_LOOPBACK_ADDRESS)
+        protected boolean allowIPv6LoopbackAddress;
+        @JsonProperty(SecureRedirectUrisEnforcerExecutorFactory.ALLOW_HTTP_SCHEME)
+        protected boolean allowHttpScheme;
+        @JsonProperty(SecureRedirectUrisEnforcerExecutorFactory.ALLOW_WILDCARD_CONTEXT_PATH)
+        protected boolean allowWildcardContextPath;
+        @JsonProperty(SecureRedirectUrisEnforcerExecutorFactory.ALLOW_PERMITTED_DOMAINS)
+        protected List<String> allowPermittedDomains = Collections.emptyList();
+        @JsonProperty(SecureRedirectUrisEnforcerExecutorFactory.OAUTH_2_1_COMPLIANT)
+        protected boolean oauth2_1complient;
+        @JsonProperty(SecureRedirectUrisEnforcerExecutorFactory.ALLOW_OPEN_REDIRECT)
+        protected boolean allowOpenRedirect;
+
+        public boolean isAllowPrivateUseUriScheme() {
+            return allowPrivateUseUriScheme;
+        }
+
+        public void setAllowPrivateUseUriScheme(boolean allowPrivateUseUriScheme) {
+            this.allowPrivateUseUriScheme = allowPrivateUseUriScheme;
+        }
+
+        public boolean isAllowIPv4LoopbackAddress() {
+            return allowIPv4LoopbackAddress;
+        }
+
+        public void setAllowIPv4LoopbackAddress(boolean allowIPv4LoopbackAddress) {
+            this.allowIPv4LoopbackAddress = allowIPv4LoopbackAddress;
+        }
+
+        public boolean isAllowIPv6LoopbackAddress() {
+            return allowIPv6LoopbackAddress;
+        }
+
+        public void setAllowIPv6LoopbackAddress(boolean allowIPv6LoopbackAddress) {
+            this.allowIPv6LoopbackAddress = allowIPv6LoopbackAddress;
+        }
+
+        public boolean isAllowHttpScheme() {
+            return allowHttpScheme;
+        }
+
+        public void setAllowHttpScheme(boolean allowHttpScheme) {
+            this.allowHttpScheme = allowHttpScheme;
+        }
+
+        public boolean isAllowWildcardContextPath() {
+            return allowWildcardContextPath;
+        }
+
+        public void setAllowWildcardContextPath(boolean allowWildcardContextPath) {
+            this.allowWildcardContextPath = allowWildcardContextPath;
+        }
+
+        public List<String> getAllowPermittedDomains() {
+            return allowPermittedDomains;
+        }
+ 
+        public void setAllowPermittedDomains(List<String> permittedDomains) {
+            this.allowPermittedDomains = permittedDomains;
+        }
```

Comment:
I think that defining domains in the executor is not very useful. Because the allowed domains in the executor are not managed later in the `RedirectUtils` verification and the URI will be declared as invalid at that point.

Maybe the idea would be adding the hostname wildcard in RedirectUtils verification and here just some check to allow those type of valid redirects or not (similar to wildcards in path option, we could even use the same option). But, for the moment, I would just remove this now and work on adding hostname wildcards as a separate issue if highly demanded in the issues.

Reply:
I am not sure I understand this suggestion. IMO it is useful to have support for "allowed permitted domains" as it allows administrator to specify for instance that all registered clients must use only domains like `example.org` when they try to register. It is also mentioned in the original design discussion https://github.com/keycloak/keycloak/discussions/9278#discussioncomment-3606032 .

Reply:
Ah! OK! So the idea is that all URLs (valid redirect URIs included) are from `example.com` for example. Although later the passed `redirect_uri` should match the configured valid redirect URIs and can fail if none matches with the passed one.
Then @tnorimat forget about this one, please close it as resolved.